cpp/src/arrow/json/options.h

```diff
@@ -0,0 +1,59 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_JSON_OPTIONS_H
+#define ARROW_JSON_OPTIONS_H
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+#include "arrow/type.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class DataType;
+
+namespace json {
+
+struct ARROW_EXPORT ParseOptions {
+  // Parsing options
+
+  // Optional explicit schema (no type inference, ignores other fields)
+  std::shared_ptr<Schema> explicit_schema;
+
+  static ParseOptions Defaults();
+};
+
+struct ARROW_EXPORT ReadOptions {
+  // Reader options
+
+  // Whether to use the global CPU thread pool
+  bool use_threads = true;
+  // Block size we request from the IO layer; also determines the size of
+  // chunks when use_threads is true
+  int32_t block_size = 1 << 20;  // 1 MB
```

Comment:
I think it would make more sense to partition on number of row as opposed to bytes. A lot more friendly to make concurrent parsing.

Reply:
Would it though? Depending on the size of lines, a block could be 500KB in some cases or 50MB in others.

Reply:
It would be nice, but this is use at read time (this is the size of blocks loaded from a file) before any parsing has been done, so we don't know where the row boundaries are. Finding those boundaries is what Chunker is for (which I haven't implemented yet but CSV has)

Reply:
@wesm while this is possible, in my time working the ad tech, this was the exception more than the norm. Byte addressable chunking also suffer from this problem, it only takes a huge line on one of the boundary.

Also I think it makes sense to dictate a size of chunks, especially if you carefully chose a size that rounds to a number of pages, say 8,16,32k. And in that case, partitioning on the number of bytes require a re-shuffling to have chunks of equal size.

Reply:
I think the concern is a bit hypothetical / low priority for the moment. Whether chunking is based on size or number of rows is not a detail that should impact the design of the rest of the code, so we don't need to burden ourselves with it right now. 

In CSV files I don't think chunking based on number of rows is always the best option. I've seen many files with 1000 columns or more in the wild. If your goal is eventually to transport the data on the wire (e.g. with gRPC), then you will end up wanting to split datasets into chunks that are around a certain size (8MB or so) anyway

Reply:
One concern with number of rows chunking is that chunking becomes more expensive, and may become a bottleneck because it's single-threaded. It's probably possible to optimize it (e.g. using masks and perhaps SIMD ops), but it's not as easy as having a fast number of bytes chunker.

Note that chunking really has two different goals here:
- the primary goal is to allow efficient multi-threaded ingestion of CSV or JSON data
- the secondary goal (which is more of a nice-to-have) is to have a chunked output with chunks of a reasonable size